### PR TITLE
shutdown: Send EXIT_STATUS before final sync

### PR DESCRIPTION
**Pull request by daandemeyer**
There's a race condition where the EXIT_STATUS= message we send just before shutting down the VM doesn't arrive on the host, presumably because the VM is shut down before the kernel has had a chance to forward the message to the host.

Since there's no obvious way to wait until the message has been flushed to the host, let's send the message before we execute the final sync() instead of after executing the final sync(). In my testing, this seems to either guarantee the message is sent or introduces sufficient delay that the kernel always has time to flush its socket buffers to the host.

### CODE REVIEW

**Comment by vcaputo** [comment]
Hm, interesting edge case.  Relying on a delay side-effect is pretty gross though, so my knee-jerk reaction is to NAK this.

Maybe a more correct solution would be polling the underlying socket's send queue size until it drains to zero?  sd-notify could have a synchronous flush for this purpose, built upon ioctl + SIOCOUTQ?
```
$ grep SIOCOUTQ /usr/include/linux/sockios.h | head -1
#define SIOCOUTQ        TIOCOUTQ        /* output queue size (not sent + not acked) */
```

**Comment by DaanDeMeyer** [comment]
@vcaputo I tried that and that ioctl is not supported for AF_VSOCK.

**Comment by vcaputo** [comment]
> @vcaputo I tried that and that ioctl is not supported for AF_VSOCK.

VSOCK? oh, I thought these were AF_UNIX...

**Comment by DaanDeMeyer** [comment]
> Edit: glancing at the kernel source for af_vsock.c it does appear to handle shutdown(), maybe that can serve this purpose if it's done synchronously with notifying the receiving side?

Tried that as well, didn't work either

File: src/shutdown/shutdown.c [hunk 1]
```diff
@@ -574,6 +574,13 @@ int main(int argc, char *argv[]) {
                           need_dm_detach ? " DM devices," : "",
                           need_md_detach ? " MD devices," : "");
 
+        /* This is primarily useful when running systemd in a VM, as it provides the user running the VM with
+         * a mechanism to pick up systemd's exit status in the VM. Note that we execute this before the final
+         * sync(), since otherwise we might shut down the VM before the AF_VSOCK buffers have been flushed.
+         * While sync() doesn't guarantee our message will arrive, it does seem to take long enough to make
+         * sure the message arrives on the listening socket. */
+        (void) sd_notifyf(0, "EXIT_STATUS=%i", arg_exit_code);
```
Comment (poettering) [inline src/shutdown/shutdown.c]: hmm, why not do this even earlier? is there any reason not to do this as one of the first things when we enter main()?

Comment (daandemeyer) [reply src/shutdown/shutdown.c]: The host will receive the message earlier. That's the only difference really. Moved it up so we do this right after parsing all the arguments.

**Pull request merged**